src/Adaptivity/AdaptedDiscreteModels.jl

```diff
@@ -0,0 +1,281 @@
+
+"""
+  AdaptedDiscreteModel
+
+  `DiscreteModel` created by refining/coarsening another `DiscreteModel`. 
+  The refinement/coarsening hierarchy can be traced backwards by following the 
+  `parent` pointer chain. This allows the transfer of dofs 
+  between `FESpaces` defined on this model and its ancestors.
+
+"""
+struct AdaptedDiscreteModel{Dc,Dp,A<:DiscreteModel{Dc,Dp},B<:DiscreteModel{Dc,Dp},C<:AdaptivityGlue} <: DiscreteModel{Dc,Dp}
+  model  ::A
+  parent ::B
+  glue   ::C
+
+  function AdaptedDiscreteModel(model::DiscreteModel{Dc,Dp},parent,glue) where {Dc,Dp}
+    @check !isa(model,AdaptedDiscreteModel)
+    A = typeof(model)
+    B = typeof(parent)
+    C = typeof(glue)
+    return new{Dc,Dp,A,B,C}(model,parent,glue)
+  end
+end
+
+# DiscreteModel API
+Geometry.get_grid(model::AdaptedDiscreteModel)          = get_grid(model.model)
+Geometry.get_grid_topology(model::AdaptedDiscreteModel) = get_grid_topology(model.model)
+Geometry.get_face_labeling(model::AdaptedDiscreteModel) = get_face_labeling(model.model)
+
+# Other getters
+get_model(model::AdaptedDiscreteModel)  = model.model
+get_parent(model::AdaptedDiscreteModel{Dc,Dp,A,<:AdaptedDiscreteModel}) where {Dc,Dp,A} = get_model(model.parent)
```

Comment:
If am not wrong, here we have an implicit post-condition on the `get_parent` function. 

Namely, the type of the output model cannot be `AdaptedDiscreteModel`. 

Should we add a test that explicitly check this post-condition? I think we dont currently have it.

Reply:
It's not necessary, this is true by construction: 
- If `.parent` is not an `AdaptedDiscreteModel`, there is nothing to check. 
- If `.parent` is an `AdaptedDiscreteModel`, then `.parent.model` is, by construction, not and `AdaptedDiscreteModel`. Therefore there is also nothing to check. The check is done in the constructor method, where we ensure `.model` is not `AdaptedDiscreteModel`.

Reply:
Yes, sure. I am thinking on what might happen in the future if one developer modifies the method such that, by mistake, it returns `AdaptedDiscreteModel`. Not a big deal anyway ...